Kuadrant/multicluster-gateway-controller
# #105: Implement DNS Health Check reconciliation

## Description

Add health check spec to DNSPolicy and implement reconciliation of the health checks on Route 53

## Verification steps

In order to verify, run the controller locally alongside the example application to ensure endpoints are added. Then, create the example DNSPolicy and verify that the status is updated and the health check is created correctly

1. Run the local setup

    ```sh
    MCTC_WORKLOAD_CLUSTERS_COUNT=1 make local-setup
    ```

1. Run the controller

    ```sh
    (export $(cat ./controller-config.env | xargs) && export $(cat ./aws-credentials.env | xargs) && make build-controller install run-controller)   
    ```

1. Create the gateway class and the tenant namespace

    ```sh
    kubectl apply -f config/samples/gatewayclass.yaml
    kubectl create namespace mctc-tenant
    ```

1. In another terminal, target the workload cluster, and run the syncer

    ```sh
    export KUBECONFIG=./tmp/kubeconfigs/mctc-workload-1.kubeconfig
    make kind-load-syncer
    make deploy-syncer

    kubectl wait --for=condition=Available deployment sync-agent -n mctc-system
    ```

1. In the terminal that targets the control plane, create the gateway

    ```sh
    kubectl apply -n mctc-tenant -f gateway.yaml
    ```

1. In the terminal that targets the workload cluster, deploy the application

    ```sh
    kubectl apply -f application.yaml
    ```

1. In the terminal that targets the control plane, create the DNSPolicy

    ```sh
    kubectl apply -n mctc-tenant -f config/samples/dnspolicy.yaml
    ```

1. The controller will reconcile the DNSPolicy and update the status with the `Created` condition in the health check section

    ```sh
    kubectl get dnspolicy example-dnspolicy -n mctc-tenant -o yaml
    ```

1. In the health check section of Route 53, in the AWS console, the new health check will be visible

    ![image](https://user-images.githubusercontent.com/12643415/233326747-eec2c014-1892-4e1f-864b-c5e4fb0cce89.png)

1. Verify in the AWS console that the DNS record for the app references the newly created health check
